org.eclipse.lsp4e/src/org/eclipse/lsp4e/operations/hover/LSPTextHover.java

```diff
@@ -200,10 +202,12 @@ private void initiateHoverRequest(@NonNull ITextViewer viewer, int offset) {
 		this.lastViewer = viewer;
 		try {
 			HoverParams params = LSPEclipseUtils.toHoverParams(offset, document);
+			this.canceller.cancel();
+			this.canceller = new Canceller();
 
 			this.request = LanguageServers.forDocument(document)
 				.withCapability(ServerCapabilities::getHoverProvider)
-				.collectAll(server -> server.getTextDocumentService().hover(params));
+				.collectAll(canceller.wrap(server -> server.getTextDocumentService().hover(params)));
```

Comment:
According to our recent discussions, we could do away with the timeout of the get operation if we can cancel, right @mickaelistria? If this is so I believe we can get rid of the timeout in the method `getHoverInfo` in line 73, because it is run on a different thread than the UI thread, see org.eclipse.jface.text.TextViewerHoverManager.computeInformation(). Do you agree?

Reply:
>  we could do away with the timeout of the get operation if we can cancel, right @mickaelistria? 

For cases where the UI is not frozen, yes; but here it's not the case, I believe the the hover request (via Eclipse API) does happen in UI Thread. If it is so then the hover strategy does need to be reworked in Platform and/or LSP4E to become non-UI blocking. Then we can consider removing the timeout.

Reply:
For the case I tested on my eclipse, it does not happen on the UI thread. I put a breakpoint on line 73 where I propose to remove the timeout, and the UI was not froozen.

Also the code I reference is quite explicit about it in org.eclipse.jface.text.TextViewerHoverManager.computeInformation()

```
		fThread= new Thread("Text Viewer Hover Presenter") { //$NON-NLS-1$
			@Override
			public void run() {
				// http://bugs.eclipse.org/bugs/show_bug.cgi?id=17693
				boolean hasFinished= false;
				try {
					if (fThread != null) {
						Object information;
						try {
							if (hover instanceof ITextHoverExtension2)
								information= ((ITextHoverExtension2)hover).getHoverInfo2(fTextViewer, region);
							else
								information= hover.getHoverInfo(fTextViewer, region);
```

so at that point, the code does not run in the UI thread. Maybe there are other consumers that do not create a new thread, so we could check if our thread is the UI one or not, and depending on that put the timeout. Does it look like a plan?

I am very keen removing the timeout because we have complex DSLs, and we actually run into timeouts, so it is a real problem us.

Reply:
OK, you're right. So yes, proper support for cancelling would make it OK to remove timeout in hover.